File: app/src/main/java/com/exemple/habitapp/HabitWidgetProvider.java
package com.exemple.habitapp;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.widget.RemoteViews;

public class HabitWidgetProvider extends AppWidgetProvider {

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        for (int appWidgetId : appWidgetIds) {
            updateWidget(context, appWidgetManager, appWidgetId);
        }
    }

    public static void updateAll(Context context) {
        AppWidgetManager manager = AppWidgetManager.getInstance(context);
        ComponentName component = new ComponentName(context, HabitWidgetProvider.class);
        int[] ids = manager.getAppWidgetIds(component);
        for (int id : ids) {
            updateWidget(context, manager, id);
        }
    }

    private static void updateWidget(Context context, AppWidgetManager manager, int appWidgetId) {
        SharedPreferences prefs = context.getSharedPreferences("habit_data", Context.MODE_PRIVATE);
        HabitStore.ensureToday(prefs);
        HabitStore.saveTodaySnapshot(prefs);

        int score = HabitStore.getTodayScore(prefs);
        int agua = HabitStore.getAguaMl(prefs);
        int metaAgua = HabitStore.getMetaAguaMl(prefs);
        int foco = prefs.getInt("estudos_concluidos_min", 0);
        int metaFoco = prefs.getInt("meta_estudos_min", 60);
        int streak = HabitStore.getStreak(prefs);
        int missingWater = Math.max(0, metaAgua - agua);
        int missingFocus = Math.max(0, metaFoco - foco);

        RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.app_widget_habit);
        views.setTextViewText(R.id.widgetTitle, "HabitApp | " + streak + "d streak");
        views.setTextViewText(R.id.widgetScore, "Score " + score + "%");
        views.setTextViewText(R.id.widgetDetails, nextWidgetText(missingWater, missingFocus, agua, metaAgua, foco, metaFoco));
        views.setProgressBar(R.id.widgetProgress, 100, score, false);
        views.setOnClickPendingIntent(R.id.widgetRoot, openAppIntent(context, R.id.home, 10));
        views.setOnClickPendingIntent(R.id.widgetAddWater, widgetAction(context, WidgetActionReceiver.ACTION_ADD_WATER, 20));
        views.setOnClickPendingIntent(R.id.widgetOpenFocus, openAppIntent(context, R.id.estudos, 30));
        manager.updateAppWidget(appWidgetId, views);
    }

    private static String nextWidgetText(int missingWater, int missingFocus, int agua, int metaAgua, int foco, int metaFoco) {
        if (missingWater > 0) return "Proximo: +" + Math.min(250, missingWater) + " ml | agua " + agua + "/" + metaAgua + " ml";
        if (missingFocus > 0) return "Proximo: foco 15 min | foco " + foco + "/" + metaFoco + " min";
        return "Dia encaminhado. Abra o app para ver conquistas.";
    }

    private static PendingIntent openAppIntent(Context context, int destination, int requestCode) {
        Intent intent = new Intent(context, MainActivity.class);
        intent.putExtra("nav_target", destination);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
        return PendingIntent.getActivity(
                context,
                requestCode,
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT | NotificationHelper.immutableFlag()
        );
    }

    private static PendingIntent widgetAction(Context context, String action, int requestCode) {
        Intent intent = new Intent(context, WidgetActionReceiver.class);
        intent.setAction(action);
        return PendingIntent.getBroadcast(
                context,
                requestCode,
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT | NotificationHelper.immutableFlag()
        );
    }
}

File: app/src/main/java/com/exemple/habitapp/widgets/HabitWidgetProvider.kt
package com.exemple.habitapp.widgets

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.widget.RemoteViews
import com.exemple.habitapp.MainActivity
import com.exemple.habitapp.R
import com.exemple.habitapp.data.HabitRepository
import com.exemple.habitapp.ui.navigation.HabitRoute

class HabitWidgetProvider : AppWidgetProvider() {
    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        appWidgetIds.forEach { updateWidget(context, appWidgetManager, it) }
    }

    companion object {
        fun updateAll(context: Context) {
            val manager = AppWidgetManager.getInstance(context)
            val component = ComponentName(context, HabitWidgetProvider::class.java)
            manager.getAppWidgetIds(component).forEach { updateWidget(context, manager, it) }
        }

        private fun updateWidget(context: Context, manager: AppWidgetManager, appWidgetId: Int) {
            val state = HabitRepository(context).snapshot()
            val missingWater = (state.waterGoalMl - state.waterMl).coerceAtLeast(0)
            val missingFocus = (state.focusGoalMinutes - state.focusMinutes).coerceAtLeast(0)
            val details = when {
                missingWater > 0 -> "Próximo: +${missingWater.coerceAtMost(250)} ml | água ${state.waterMl}/${state.waterGoalMl} ml"
                missingFocus > 0 -> "Próximo: foco 15 min | foco ${state.focusMinutes}/${state.focusGoalMinutes} min"
                else -> "Dia encaminhado. Abra o app para ver conquistas."
            }

            val views = RemoteViews(context.packageName, R.layout.app_widget_habit).apply {
                setTextViewText(R.id.widgetTitle, "HabitApp | ${state.streak}d streak")
                setTextViewText(R.id.widgetScore, "Score ${state.score}%")
                setTextViewText(R.id.widgetDetails, details)
                setProgressBar(R.id.widgetProgress, 100, state.score, false)
                setOnClickPendingIntent(R.id.widgetRoot, openAppIntent(context, HabitRoute.Home.route, 10))
                setOnClickPendingIntent(R.id.widgetAddWater, widgetAction(context, WidgetActionReceiver.ACTION_ADD_WATER, 20))
                setOnClickPendingIntent(R.id.widgetOpenFocus, openAppIntent(context, HabitRoute.Focus.route, 30))
            }
            manager.updateAppWidget(appWidgetId, views)
        }

        private fun openAppIntent(context: Context, route: String, requestCode: Int): PendingIntent {
            val intent = Intent(context, MainActivity::class.java)
                .putExtra(MainActivity.EXTRA_NAV_ROUTE, route)
                .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP)
            return PendingIntent.getActivity(
                context,
                requestCode,
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
            )
        }

        private fun widgetAction(context: Context, action: String, requestCode: Int): PendingIntent {
            val intent = Intent(context, WidgetActionReceiver::class.java).setAction(action)
            return PendingIntent.getBroadcast(
                context,
                requestCode,
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
            )
        }
    }
}
